net: lwm2m: Fix issues reported by undefined behavior sanitizer

rlubos opened the pull request:
For details, please check individual commits.

rlubos committed:
net: lwm2m: registry: Validate value pointer before use

Check if value pointer is not NULL before passing it to memcmp() inside
lwm2m_engine_set(). As the function actually expects that the value
pointer can be NULL in case resource value is cleared (there is a test
case for such behavior), validate that len value is actually 0 if NULL
value is provided, to avoid unexpected behavior in other parts of the
function.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: json: Check for int64_t overflow when parsing integer

Verify if the integer value being parsed does not overflow int64_t type
and report an error in such cases.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: plain_text: Check for int64_t overflow when parsing integer

Verify if the integer value being parsed does not overflow int64_t type
and report an error in such cases.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: oma_tlv: Fix UBSAN warnings about signed integers

* Simplify the logic of the get_number() function to address the corner
  cases reported by UBSAN regarding byte-swapping signed integer values.

  The existing logic was actually only valid for little-endian systems,
  as it expected that the bytes written from the packet buffer will be
  stored at the beginning of the int64_t memory, plus the actual
  byte-swapping with signed integer casts inside was hard to follow.
  Switch to a plain uint8_t buffer for integer readout, and use
  dedicated system function to convert the big-endian data in the buffer
  into unsigned integer in the system endianness, followed by the final
  cast to a signed value.

* Add explicit cast to uint32_t in put_objlnk() to prevent warning about
  not-fitting integer after byte shift, and update the result type to
  uint32_t as well.

* Switch to buffer with sys_put_be16/32/64 when writing integers due to
  similar warnings about byte-swapping signed values.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: senml_cbor: Verify data len before calling memcpy

Check if there actually is data to copy before calling memcpy() to
prevent potentially calling memcpy() with NULL value pointer.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

SeppoTakalo reviewed (dismissed):
Looks good otherwise than I think the int64_t conversion on JSON might still roll-over.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rw_json.c:
```diff
@@ -663,17 +664,18 @@ static int read_int(struct lwm2m_input_context *in, int64_t *value,
+			temp = temp * 10ULL + (c - '0');
+			if (temp > (INT64_MAX + (neg ? 1ULL : 0ULL))) {
+				return -EINVAL;
+			}
```
Good point, I've updated the check for json and plain_text.

rlubos replied:
On the second though, the original proposal was fine, the roll-over is not a concern as we update `uint64_t` variable and then compare against int64_t limits.

rlubos committed:
tests: net: lwm2m: client: Fix stack memory corruptions

LwM2M client context was defined on stack in the test function, however
it could still be in use when the test ended, as the actual LwM2M
teardown took place in a common "after" test function. In result, the
Lwm2M context content could be corrupted.

Additionally, increase the system work queue stack size, as the stack
overflowed.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

kartben commented on subsys/net/lib/lwm2m/lwm2m_rw_json.c:
```diff
@@ -663,17 +664,19 @@ static int read_int(struct lwm2m_input_context *in, int64_t *value,
 			neg = true;
 		} else if (isdigit(c) != 0) {
-			*value = *value * 10 + (c - '0');
+			if (temp > INT64_MAX / 10ULL) {
+				return -EINVAL;
+			}
+
+			temp = temp * 10ULL + (c - '0');
```
INT64_MAX = `9223372036854775807` so you're basically allowing `9223372036854775808` and `9223372036854775809` to sneak in (confirmed by testing)

also make sure that when you fix it you still allow `-9223372036854775808` to be a valid input. Currently if the highlighted code were to work as I think you wanted to, it would have not allowed for `9223372036854775808`

rlubos committed:
tests: net: lwm2m: content_json: Check for overflows

Add checks for int64_t overflows when parsing.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
tests: net: lwm2m: content_plain_text: Check for overflows

Add checks for int64_t overflows when parsing.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>
